File: 1528.shuffle-string.py
#
# @lc app=leetcode id=1528 lang=python3
#
# [1528] Shuffle String
#

# @lc code=start
class Solution:
    def restoreString(self, s: str, indices: List[int]) -> str:
        width = len(s)
        s_ls = list(s)
        pos = 0
        while pos < width:
            if indices[pos] == pos:
                pos += 1
            else:
                s_ls[pos], s_ls[indices[pos]] = s_ls[indices[pos]], s_ls[pos]
                swap_pos = indices[pos]
                indices[pos], indices[swap_pos] = indices[swap_pos], indices[pos]
        return "".join(s_ls)
# @lc code=end

File: 1528.shuffle-string.cpp
/*
 * @lc app=leetcode id=1528 lang=cpp
 *
 * [1528] Shuffle String
 */
#include <string>
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    string restoreString(string s, vector<int> &indices)
    {
        int width = s.size();
        int pos = 0;
        while (pos != width)
        {
            if (indices[pos] == pos)
            {
                ++pos;
            }
            else
            {
                swap(s[pos], s[indices[pos]]);
                swap(indices[pos], indices[indices[pos]]);
            }
        }
        return s;
    }
};
// @lc code=end
